pymeasure/instruments/keithley/keithley6517b.py

```diff
@@ -0,0 +1,343 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2019 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+import time
+import re
+
+import numpy as np
+
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import truncated_range, strict_discrete_set
+from .buffer import KeithleyBuffer
+
+# Setup logging
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+class Keithley6517B(Instrument, KeithleyBuffer):
+    """ Represents the Keithely 6517B ElectroMeter and provides a
+    high-level interface for interacting with the instrument.
+
+    .. code-block:: python
+
+        keithley = Keithley6517B("GPIB::1")
+
+        keithley.apply_voltage()                # Sets up to source current
+        keithley.source_voltage_range = 200     # Sets the source voltage range to 200 V
+        keithley.source_voltage = 20            # Sets the source voltage to 20 V
+        keithley.enable_source()                # Enables the source output
+
+        keithley.measure_resistance()           # Sets up to measure resistance
+
+        keithley.ramp_to_voltage(50)            # Ramps the voltage to 50 V
+        print(keithley.resistance)              # Prints the resistance in Ohms
+
+        keithley.shutdown()                     # Ramps the voltage to 0 V and disables output
+
+    """
+
+    source_enabled = Instrument.measurement("OUTPUT?",
+        """ Reads a boolean value that is True if the source is enabled. """,
+        cast=bool
+    )
+
+
+    def extract_value(result):
```

Comment:
This function seems that it could be generalized out of this class. Are there other places it can be reused? I would suggest also providing more details in the docs.

Reply:
I think we have achieved what we can here, now.